editoast/src/views/timetable/train_schedule.rs

```diff
@@ -564,6 +574,147 @@ async fn occupancy_blocks(
+            |missing| TrainScheduleError::BatchTrainScheduleNotFound {
+                number: missing.len(),
+            },
+        )
+        .await?;
+
+    let operational_point = OperationalPointModel::retrieve_real_or_fail(
+        db_pool.get().await?,
+        (infra_id, op_id.clone()),
+        || TrainScheduleError::OperationalPointNotFound { op_id },
+    )
+    .await?
+    .schema;
+
+    // Retrieve simulations
+    let simulations_result = train_simulation_batch(
+        &mut db_pool.get().await?,
+        valkey,
+        core_client,
+        &train_schedules,
+        &infra,
+        electrical_profile_set_id,
+    )
+    .await?;
+    let operational_point_track_offsets = operational_point.track_offset();
+
+    for ((simulation, pathfinding), train_schedule) in
+        simulations_result.iter().zip(train_schedules)
+    {
+        let track_ranges = match pathfinding {
+            PathfindingResult::Success(pathfinding_result_success) => {
+                &pathfinding_result_success.track_section_ranges
+            }
+            PathfindingResult::Failure(_) => continue,
+        };
+        let mut abs_offset = 0;
+        for track_range in track_ranges {
+            let track_section = track_range.clone().track_section.0;
+            let range_length = track_range.end - track_range.begin;
+            if operational_point_track_offsets
+                .iter()
+                .any(|to| to.track == track_range.track_section)
+            {
+                let abs_begin = abs_offset + track_range.begin;
+                let abs_end = abs_offset + track_range.end;
+                abs_offset += range_length;
+                let (positions, times) = match simulation {
+                    Response::Success { final_output, .. } => (
+                        final_output.report_train.positions.clone(),
+                        final_output.report_train.times.clone(),
+                    ),
+                    _ => continue,
+                };
+
+                let time_begin = interpolated_time_at_position(&positions, &times, abs_begin);
+                let time_end = interpolated_time_at_position(&positions, &times, abs_end);
+
+                if let (Some(time_begin), Some(time_end)) = (time_begin, time_end) {
+                    let duration = Duration::milliseconds((time_end - time_begin) as i64);
+                    result
+                        .entry(track_section.clone())
+                        .or_default()
+                        .push(TrackOccupancy {
+                            time_begin: train_schedule.start_time
+                                + Duration::milliseconds(time_begin.try_into().unwrap()),
+                            duration: duration.try_into().unwrap(),
+                            train_schedule_id: train_schedule.id,
+                        });
+                }
+            } else {
+                abs_offset += range_length;
+            }
+        }
+    }
```

Comment:
The algorithm should be extracted into a local function and *thoroughly documented* so that the next reader won't have to search for the closed issue to understand what's going on.